sdk/src/main/java/io/opentelemetry/sdk/trace/data/DelegatingSpanData.java

```diff
@@ -0,0 +1,257 @@
+package io.opentelemetry.sdk.trace.data;
+
+import static java.util.Objects.requireNonNull;
+
+import io.opentelemetry.common.ReadableAttributes;
+import io.opentelemetry.sdk.common.InstrumentationLibraryInfo;
+import io.opentelemetry.sdk.resources.Resource;
+import io.opentelemetry.trace.Span.Kind;
+import io.opentelemetry.trace.SpanId;
+import io.opentelemetry.trace.Status;
+import io.opentelemetry.trace.TraceFlags;
+import io.opentelemetry.trace.TraceId;
+import io.opentelemetry.trace.TraceState;
+import java.util.List;
+
+/**
+ * A {@link SpanData} which delegates all methods to another {@link SpanData}. Extend this class
+ * to modify the {@link SpanData} that will be exported, for example by creating a delegating
+ * {@link io.opentelemetry.sdk.trace.export.SpanExporter} which wraps
+ * {@link SpanData} with a custom implementation.
+ *
+ * <pre>{@code
+ *
+ * SpanDataWithClientType extends DelegatingSpanData {
+ *
+ *   private final ReadableAttributes attributes;
+ *
+ *   SpanDataWithClientType(SpanData delegate) {
+ *     super(delegate);
+ *     String clientType = ClientConfig.parseUserAgent(
+ *       delegate.getAttributes().get(SemanticAttributes.HTTP_USER_AGENT).getStringValue());
+ *     Attributes.Builder newAttributes = Attributes.newBuilder();
+ *     delegate.getAttributes().forEach(newAttributes::setAttribute);
+ *     newAttributes.setAttribute("client_type", clientType);
+ *     attributes = newAttributes.build();
```

Comment:
yeah, this was what I was thinking, @iNikem . It's not super pretty, but could be made much nicer with a `toBuilder` on the Attributes class, like we have on Labels.